File: apps/haskell/part1/test/ImmutabilitySpec.hs
{-|
Module      : ImmutabilitySpec
Description : Tests for Chapter 1 - Immutability and Data Transformation
-}
module ImmutabilitySpec (spec) where

import Test.Hspec
import Test.QuickCheck
import Immutability

spec :: Spec
spec = do
    describe "Person - Basic Immutability" $ do
        it "creates a person with name and age" $ do
            let person = Person "田中" 30
            personName person `shouldBe` "田中"
            personAge person `shouldBe` 30

        it "updateAge returns a new person with updated age" $ do
            let original = Person "田中" 30
            let updated = updateAge original 31
            personAge updated `shouldBe` 31

        it "updateAge does not modify the original person" $ do
            let original = Person "田中" 30
            let _ = updateAge original 31
            personAge original `shouldBe` 30

    describe "Team - Structural Sharing" $ do
        it "creates a team with members" $ do
            let team = Team "開発チーム" [Member "田中" Developer]
            teamName team `shouldBe` "開発チーム"
            length (teamMembers team) `shouldBe` 1

        it "addMember returns a new team with additional member" $ do
            let team = Team "開発チーム" [Member "田中" Developer]
            let newTeam = addMember team (Member "鈴木" Designer)
            length (teamMembers newTeam) `shouldBe` 2

        it "addMember does not modify the original team" $ do
            let team = Team "開発チーム" [Member "田中" Developer]
            let _ = addMember team (Member "鈴木" Designer)
            length (teamMembers team) `shouldBe` 1

    describe "Order Processing - Data Pipeline" $ do
        it "calculates subtotal for an item" $ do
            let item = Item "商品A" 1000 2
            calculateSubtotal item `shouldBe` 2000

        it "returns correct membership discount" $ do
            membershipDiscount Gold `shouldBe` 0.1
            membershipDiscount Silver `shouldBe` 0.05
            membershipDiscount Bronze `shouldBe` 0.02
            membershipDiscount Regular `shouldBe` 0.0

        it "calculates total for an order" $ do
            let items = [ Item "商品A" 1000 2
                       , Item "商品B" 500 3
                       , Item "商品C" 2000 1
                       ]
            let customer = Customer "山田" Gold
            let order = Order items customer
            calculateTotal order `shouldBe` 5500

        it "applies membership discount" $ do
            let items = [ Item "商品A" 1000 2
                       , Item "商品B" 500 3
                       , Item "商品C" 2000 1
                       ]
            let customer = Customer "山田" Gold
            let order = Order items customer
            applyDiscount order 5500 `shouldBe` 4950.0

        it "processes order with full pipeline" $ do
            let items = [ Item "商品A" 1000 2
                       , Item "商品B" 500 3
                       , Item "商品C" 2000 1
                       ]
            let customer = Customer "山田" Gold
            let order = Order items customer
            processOrder order `shouldBe` 4950.0

    describe "Invoice - Pure Function" $ do
        it "calculates invoice with tax" $ do
            let items = [Item "商品A" 1000 1]
            let invoice = calculateInvoice items 0.1
            invoiceSubtotal invoice `shouldBe` 1000
            invoiceTax invoice `shouldBe` 100.0
            invoiceTotal invoice `shouldBe` 1100.0

    describe "History - Undo/Redo" $ do
        it "creates empty history" $ do
            let history = createHistory :: History Int
            currentState history `shouldBe` Nothing
            canUndo history `shouldBe` False
            canRedo history `shouldBe` False

        it "pushes state onto history" $ do
            let h1 = createHistory :: History Int
            let h2 = pushState h1 1
            currentState h2 `shouldBe` Just 1

        it "supports undo operation" $ do
            let h1 = createHistory :: History Int
            let h2 = pushState h1 1
            let h3 = pushState h2 2
            let h4 = undo h3
            currentState h4 `shouldBe` Just 1
            canRedo h4 `shouldBe` True

        it "supports redo operation" $ do
            let h1 = createHistory :: History Int
            let h2 = pushState h1 1
            let h3 = pushState h2 2
            let h4 = undo h3
            let h5 = redo h4
            currentState h5 `shouldBe` Just 2

        it "clears future when pushing new state after undo" $ do
            let h1 = createHistory :: History Int
            let h2 = pushState h1 1
            let h3 = pushState h2 2
            let h4 = undo h3
            let h5 = pushState h4 3
            currentState h5 `shouldBe` Just 3
            canRedo h5 `shouldBe` False

        it "undo on empty history returns same history" $ do
            let history = createHistory :: History Int
            let undone = undo history
            currentState undone `shouldBe` Nothing

        it "redo on empty future returns same history" $ do
            let h1 = createHistory :: History Int
            let h2 = pushState h1 1
            let h3 = redo h2
            currentState h3 `shouldBe` Just 1

    describe "QuickCheck Properties" $ do
        it "updateAge preserves name" $ property $
            \name (NonNegative age) (NonNegative newAge) ->
                let person = Person name age
                    updated = updateAge person newAge
                in personName updated == name

        it "calculateSubtotal is price * quantity" $ property $
            \(Positive price) (Positive qty) ->
                let item = Item "Test" price qty
                in calculateSubtotal item == price * qty

File: apps/scala/part1/src/test/scala/ImmutabilitySpec.scala
package immutability

import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

class ImmutabilitySpec extends AnyFlatSpec with Matchers:

  import Immutability.*

  // =============================================================================
  // 不変データ構造の基本
  // =============================================================================

  "updateAge" should "元のデータを変更しない" in {
    val original = Person("田中", 30)
    val updated = updateAge(original, 31)

    original.age shouldBe 30
    updated.age shouldBe 31
    updated.name shouldBe "田中"
  }

  // =============================================================================
  // 構造共有
  // =============================================================================

  "addMember" should "新しいメンバーを追加した新しいチームを返す" in {
    val team = Team("開発チーム", List(Member("田中", "developer")))
    val newTeam = addMember(team, Member("鈴木", "designer"))

    team.members.length shouldBe 1
    newTeam.members.length shouldBe 2
    newTeam.name shouldBe "開発チーム"
  }

  // =============================================================================
  // データ変換パイプライン
  // =============================================================================

  "calculateSubtotal" should "単価×数量を計算する" in {
    calculateSubtotal(Item("A", 1000, 2)) shouldBe 2000
    calculateSubtotal(Item("B", 500, 3)) shouldBe 1500
  }

  "membershipDiscount" should "会員種別に応じた割引率を返す" in {
    membershipDiscount("gold") shouldBe 0.1
    membershipDiscount("silver") shouldBe 0.05
    membershipDiscount("bronze") shouldBe 0.02
    membershipDiscount("regular") shouldBe 0.0
  }

  "calculateTotal" should "注文の合計金額を計算する" in {
    val order = Order(
      items = List(Item("A", 1000, 2), Item("B", 500, 3)),
      customer = Customer("山田", "regular")
    )
    calculateTotal(order) shouldBe 3500
  }

  "applyDiscount" should "割引後の金額を計算する" in {
    val order = Order(
      items = List.empty,
      customer = Customer("山田", "gold")
    )
    applyDiscount(order, 1000) shouldBe 900.0
  }

  "processOrder" should "注文処理全体を行う" in {
    val order = Order(
      items = List(Item("A", 1000, 2), Item("B", 500, 3)),
      customer = Customer("山田", "gold")
    )
    processOrder(order) shouldBe 3150.0
  }

  // =============================================================================
  // 副作用の分離
  // =============================================================================

  "pureCalculateTax" should "税額を計算する" in {
    pureCalculateTax(1000, 0.1) shouldBe 100.0
    pureCalculateTax(1000, 0.08) shouldBe 80.0
  }

  "calculateInvoice" should "請求書を計算する" in {
    val items = List(Item("A", 1000, 2), Item("B", 500, 1))
    val invoice = calculateInvoice(items, 0.1)

    invoice.subtotal shouldBe 2500
    invoice.tax shouldBe 250.0
    invoice.total shouldBe 2750.0
  }

  // =============================================================================
  // Undo/Redo の実装
  // =============================================================================

  "createHistory" should "空の履歴を作成する" in {
    val history = createHistory[Map[String, String]]()

    history.current shouldBe None
    history.past shouldBe List.empty
    history.future shouldBe List.empty
  }

  "pushState" should "新しい状態を追加する" in {
    val history = pushState(createHistory[Map[String, String]](), Map("text" -> "Hello"))

    history.current shouldBe Some(Map("text" -> "Hello"))
    history.past shouldBe List(None).flatten
  }

  "undo" should "直前の状態に戻す" in {
    val history = pushState(
      pushState(createHistory[Map[String, String]](), Map("text" -> "Hello")),
      Map("text" -> "Hello World")
    )
    val afterUndo = undo(history)

    afterUndo.current shouldBe Some(Map("text" -> "Hello"))
    afterUndo.future shouldBe List(Map("text" -> "Hello World"))
  }

  "redo" should "やり直しを行う" in {
    val history = redo(undo(pushState(
      pushState(createHistory[Map[String, String]](), Map("text" -> "Hello")),
      Map("text" -> "Hello World")
    )))

    history.current shouldBe Some(Map("text" -> "Hello World"))
    history.future shouldBe List.empty
  }

  "undo" should "空の履歴で undo しても安全" in {
    val history = createHistory[Map[String, String]]()
    val afterUndo = undo(history)

    afterUndo.current shouldBe None
  }

  "redo" should "future が空で redo しても安全" in {
    val history = pushState(createHistory[Map[String, String]](), Map("text" -> "Hello"))
    val afterRedo = redo(history)

    afterRedo.current shouldBe Some(Map("text" -> "Hello"))
  }

  // =============================================================================
  // トランスデューサー（Scala では関数合成）
  // =============================================================================

  "processItemsEfficiently" should "効率的にアイテムを処理する" in {
    val items = List(
      Item("A", 100, 2),   // subtotal 200 > 100 ✓
      Item("B", 50, 1),    // subtotal 50 <= 100 ✗
      Item("C", 200, 0),   // quantity 0 ✗
      Item("D", 300, 1)    // subtotal 300 > 100 ✓
    )
    processItemsEfficiently(items).length shouldBe 2
  }

  "processItemsWithIterator" should "Iterator を使用して効率的にアイテムを処理する" in {
    val items = List(
      Item("A", 100, 2),
      Item("B", 50, 1),
      Item("C", 200, 0),
      Item("D", 300, 1)
    )
    processItemsWithIterator(items).length shouldBe 2
  }
